[auto_functionalized] Support `Tensor(a!)[]?`

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145400

Summary:
This is just updating some of the checks to allow the Tensor(a!)[]? type
through.

Fixes #144072

Test Plan:
- new tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

zou3519 committed:
[auto_functionalized] Support `Tensor(a!)[]?`

Summary:
This is just updating some of the checks to allow the Tensor(a!)[]? type
through.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[auto_functionalized] Support `Tensor(a!)[]?`"


Summary:
This is just updating some of the checks to allow the Tensor(a!)[]? type
through.

Fixes #144072

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[auto_functionalized] Support `Tensor(a!)[]?`"


Summary:
This is just updating some of the checks to allow the Tensor(a!)[]? type
through.

Fixes #144072

Test Plan:
- new tests

[ghstack-poisoned]

laithsakka reviewed (approved):
looks good would it make sense to add a test in est/inductor/test_auto_functionalize.py for the previously unsupported case?

zou3519 commented:
> looks good would it make sense to add a test in est/inductor/test_auto_functionalize.py for the previously unsupported case?

The test I would add is exactly what the test case in this PR is, so maybe not? Unless you have suggestions for what else to test

zou3519 committed:
Update on "[auto_functionalized] Support `Tensor(a!)[]?`"


Summary:
This is just updating some of the checks to allow the Tensor(a!)[]? type
through.

Fixes #144072

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "rocm sev"
